### PR TITLE
Polars: length transformation

### PR DESCRIPTION
**Pull request by Shoeboxam**
implement: 
- #1538

### CODE REVIEW

**Comment by Shoeboxam** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/opendp/opendp/1540?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#1540** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1540?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#1539** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1539?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1534** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1534?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1519** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1519?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1506** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1506?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1498** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1498?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1479** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1479?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#1496](https://github.com/opendp/opendp/pull/1496) <a href="https://app.graphite.dev/github/pr/opendp/opendp/1496?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#908** <a href="https://app.graphite.dev/github/pr/opendp/opendp/908?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#913](https://github.com/opendp/opendp/pull/913) <a href="https://app.graphite.dev/github/pr/opendp/opendp/913?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#890** <a href="https://app.graphite.dev/github/pr/opendp/opendp/890?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2 other dependent PRs ([#774](https://github.com/opendp/opendp/pull/774) <a href="https://app.graphite.dev/github/pr/opendp/opendp/774?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#856](https://github.com/opendp/opendp/pull/856) <a href="https://app.graphite.dev/github/pr/opendp/opendp/856?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#829** <a href="https://app.graphite.dev/github/pr/opendp/opendp/829?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2 other dependent PRs ([#854](https://github.com/opendp/opendp/pull/854) <a href="https://app.graphite.dev/github/pr/opendp/opendp/854?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#888](https://github.com/opendp/opendp/pull/888) <a href="https://app.graphite.dev/github/pr/opendp/opendp/888?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#1515** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1515?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#819** <a href="https://app.graphite.dev/github/pr/opendp/opendp/819?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#868** <a href="https://app.graphite.dev/github/pr/opendp/opendp/868?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#864](https://github.com/opendp/opendp/pull/864) <a href="https://app.graphite.dev/github/pr/opendp/opendp/864?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#797** <a href="https://app.graphite.dev/github/pr/opendp/opendp/797?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#824](https://github.com/opendp/opendp/pull/824) <a href="https://app.graphite.dev/github/pr/opendp/opendp/824?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#847** <a href="https://app.graphite.dev/github/pr/opendp/opendp/847?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1454** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1454?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#795** <a href="https://app.graphite.dev/github/pr/opendp/opendp/795?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1167** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1167?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#769** <a href="https://app.graphite.dev/github/pr/opendp/opendp/769?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#771](https://github.com/opendp/opendp/pull/771) <a href="https://app.graphite.dev/github/pr/opendp/opendp/771?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#767** <a href="https://app.graphite.dev/github/pr/opendp/opendp/767?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#765** <a href="https://app.graphite.dev/github/pr/opendp/opendp/765?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#763** <a href="https://app.graphite.dev/github/pr/opendp/opendp/763?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#917** <a href="https://app.graphite.dev/github/pr/opendp/opendp/917?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#921](https://github.com/opendp/opendp/pull/921) <a href="https://app.graphite.dev/github/pr/opendp/opendp/921?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#916** <a href="https://app.graphite.dev/github/pr/opendp/opendp/916?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#907** <a href="https://app.graphite.dev/github/pr/opendp/opendp/907?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#874** <a href="https://app.graphite.dev/github/pr/opendp/opendp/874?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#705** <a href="https://app.graphite.dev/github/pr/opendp/opendp/705?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#754** <a href="https://app.graphite.dev/github/pr/opendp/opendp/754?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#704** <a href="https://app.graphite.dev/github/pr/opendp/opendp/704?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#703** <a href="https://app.graphite.dev/github/pr/opendp/opendp/703?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#753** <a href="https://app.graphite.dev/github/pr/opendp/opendp/753?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#702** <a href="https://app.graphite.dev/github/pr/opendp/opendp/702?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Shoeboxam and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: rust/src/transformations/make_stable_expr/expr_len/mod.rs [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+use crate::metrics::{IntDistance, LpDistance, PartitionDistance};
+use crate::traits::{InfCast, InfMul, InfSqrt, ProductOrd};
+use crate::transformations::traits::UnboundedMetric;
+use polars_plan::dsl::{len, Expr};
+use polars_plan::logical_plan::LogicalPlan;
+
+#[cfg(test)]
+mod test;
+
+/// Polars operator to sum a column in a LazyFrame
```
Comment (mccalluc) [inline rust/src/transformations/make_stable_expr/expr_len/mod.rs:13]: Is "sum" correct here?

File: rust/src/transformations/make_stable_expr/expr_len/mod.rs [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    MI: 'static + UnboundedMetric,
+    (ExprDomain, PartitionDistance<MI>): MetricSpace,
+    (ExprDomain, LpDistance<P, f64>): MetricSpace,
+{
+    let Expr::Len = expr else {
+        return fallible!(MakeTransformation, "expected len expression");
+    };
+
+    // check that we are in a context where it is ok to break row-alignment
+    input_domain.context.break_alignment()?;
```
Comment (mccalluc) [inline rust/src/transformations/make_stable_expr/expr_len/mod.rs:39]: I've seen this idiom before, but I'm not sure if we exercise it failing anywhere?

File: rust/src/transformations/make_stable_expr/expr_len/mod.rs [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    // we only care about the margin that matches the current grouping columns
+    let margin_id = input_domain.context.grouping_columns()?;
+    let public_info = (input_domain.frame_domain.margins.get(&margin_id))
+        .map(|m| m.public_info.clone())
+        .unwrap_or_default();
+
+    let norm_map = move |d_in: f64| match P {
+        1 => Ok(d_in),
+        2 => d_in.inf_sqrt(),
+        _ => return fallible!(MakeTransformation, "unsupported Lp norm"),
```
Comment (mccalluc) [inline rust/src/transformations/make_stable_expr/expr_len/mod.rs:56]: Include the requested norm in the error message?

File: rust/src/transformations/make_stable_expr/expr_len/mod.rs [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        _ => *d_in,
+    };
+
+    let stability_map = StabilityMap::new_fallible(
+        move |(l0, l1, l_inf): &(IntDistance, IntDistance, IntDistance)| {
+            let l0_p = norm_map(f64::from(*l0))?;
+            let l1_p = f64::inf_cast(pp_map(l1))?;
+            let l_inf_p = f64::inf_cast(pp_map(l_inf))?;
+
+            l1_p.total_min(l0_p.inf_mul(&l_inf_p)?)
```
Comment (mccalluc) [inline rust/src/transformations/make_stable_expr/expr_len/mod.rs:70]: Comment that restates this in more math-y language?

File: rust/src/transformations/make_stable_expr/expr_len/mod.rs [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    );
+
+    Transformation::new(
+        input_domain,
+        output_domain,
+        Function::new_fallible(
+            // in most other situations, we would use `Function::new_expr`, but we need to return a Fallible here
+            move |(lp, expr): &(LogicalPlan, Expr)| -> Fallible<(LogicalPlan, Expr)> {
+                if expr != &Expr::Wildcard {
+                    return fallible!(
```
Comment (mccalluc) [inline rust/src/transformations/make_stable_expr/expr_len/mod.rs:81]: Should a test exercise this path?

File: rust/src/transformations/make_stable_expr/expr_len/test.rs [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+#[test]
+fn test_select_make_expr_len() -> Fallible<()> {
+    let (lf_domain, lf) = get_test_data()?;
+    let expr_domain = lf_domain.aggregate(["chunk_2_bool", "cycle_5_alpha"]);
+
+    let t_sum: Transformation<_, _, _, L2Distance<f64>> =
+        len().make_stable(expr_domain, PartitionDistance(InsertDeleteDistance))?;
+    let expr_res = t_sum.invoke(&(lf.logical_plan.clone(), all()))?.1;
+    assert_eq!(expr_res, len());
```
Comment (mccalluc) [inline rust/src/transformations/make_stable_expr/expr_len/test.rs:19]: This is coming from `super::*`? Maybe drop that and add `polars_plan::dsl::len` so it's easier for the reader to know where names come from.

File: rust/src/transformations/make_stable_expr/expr_len/test.rs [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+    let sens = t_sum.map(&(4, 4, 1))?;
+    println!("sens: {:?}", sens);
+    assert_eq!(sens, 2.);
```
Comment (mccalluc) [inline rust/src/transformations/make_stable_expr/expr_len/test.rs:23]: I'd spell out `sensitivity`.

File: rust/src/transformations/make_stable_expr/expr_len/test.rs [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    // = min(sqrt(l0) * map(linf)         , map(l1))
+    // = min(sqrt(l0) * linf * max(|L|, U), l1 * max(|L|, U))
+    // = min(sqrt(10) * 1, 10)
+    // = min(3.16227, 10)
+    // = 3.16227
+
+    // that is, in the worst case, we know the sum will differ by at most 1 in 10 partitions,
+    // so the l2 distance between any two outputs on neighboring data sets is at most 3.16227
+
+    // The sensitivity is slightly higher to account for potential rounding errors.
```
Comment (mccalluc) [inline rust/src/transformations/make_stable_expr/expr_len/test.rs:67]: This explanation is helpful! I think I was hoping for something like this in the `mod.rs`.

**Review by mccalluc — COMMENTED** [review]
a couple comments

**Review by mccalluc — DISMISSED** [review]
Looks good. I think some of the former comments were on code that just isn't here at this point.

**Pull request merged**